Add support for running scenarios with locally-defined services

AshFungor opened the pull request:
Hey, @riebl!

I've realized that problem with libraries & ned folders originally was caused by the placement of script's generator function. If placed after processing scenarios, all dependencies are detected and formatted just fine. Although, I wonder if it is a good idea to put all services into one variable globally.

Special file with parameters for each scenario is a great alternative, but it seems like a significant overhead and might confuse people too much so they stop using this script completely.

What do you think?

solves: #365

AshFungor commented:
On second thought - we might just add current working directory of a running scipt (parent directory of omnetpp.ini) to ned folders. Libraries will be added globally (you cannot define targets with same names in CMake, anyway).

This should also work.

riebl commented on cmake/run_artery.py.in:
```diff
@@ -0,0 +1,69 @@
+OPP_RUN_LIBRARIES = '@opp_run_libraries@'
+
+
+for template, value in copy.copy(globals()).items():
```
well, and this safety net would also become obsolete with a config file

riebl commented on cmake/run_artery.py.in:
```diff
@@ -0,0 +1,69 @@
+
+    parser.add_argument('--all', action='store_true', help='sets runall mode')
+    parser.add_argument('-b', '--batchsize', dest='batch', action='store')
+    parser.add_argument('-j', '--jobs', dest='jobs', action='store')
```
We can even add an option here to specify another config file location :-)

riebl commented on cmake/run_artery.py.in:
```diff
@@ -0,0 +1,69 @@
+
+    print('running command: ', ' '.join(cmd))
+
+    sys.exit(subprocess.run(
```
Uhm, I prefer to have this separated:
```
proc = subprocess.run(...)
sys.exit(proc.returncode)
```

riebl reviewed (commented):
I like this draft and it avoids my too complicated over-engineering :-)

AshFungor commented on CMakeLists.txt:
```diff
@@ -95,3 +95,12 @@ option(VSCODE_LAUNCH_INTEGRATION "Generate VS Code configuration for debugging A
+
+get_property(is_multi_config GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+if(NOT is_multi_config)
+    generate_run_script(TARGET artery FILE ${PROJECT_BINARY_DIR}/${CMAKE_CFG_INTDIR}/run_artery.py)
```
Yeah, I got rid of it :)

AshFungor commented on CMakeLists.txt:
```diff
@@ -95,3 +95,12 @@ option(VSCODE_LAUNCH_INTEGRATION "Generate VS Code configuration for debugging A
 endif()
+
+get_property(is_multi_config GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+if(NOT is_multi_config)
```
Done!

AshFungor commented:
Hey, @riebl!

Apart from your suggestions I also added .flake8 formatting settings (considering we now have extensive amount of python code, some formatting seems reasonable).

My only problem with this PR are install targets. Not all libraries & ned folders are getting installed, and with multi-config generators I cannot think of a way to generate options in uniform way. I would like to know your thoughts on this :)

I removed installation of runner script and launch configs temporary.

riebl commented on cmake/run-artery-config.ini.in:
```diff
@@ -0,0 +1,5 @@
+[General]
```
We can think about splitting these variables (I know, they are just four for now…) in logical groups, e.g. [Executables] and [Artery]. Again, this is just an idea for the moment.

AshFungor replied:
To be honest I did exactly this when I first designed this config and thought that it might be just a little too much :)

riebl commented on tools/build.py:
```diff
@@ -49,6 +49,7 @@ class Config:
     cores: int = multiprocessing.cpu_count()
     profile: typing.Union[pathlib.Path, str] = 'default'
     generator: str = 'Ninja'
+    is_generator_multiconfig: bool = False
```
I think it is not a good idea to add this option because it can be set wrong accidentally by less experienced CMake users.
Instead, we can look up if the used generator is "multi config" via https://cmake.org/cmake/help/v3.31/manual/cmake-file-api.7.html#v1-reply-index-file

AshFungor replied:
I think the most reasonable option here is to just drop support for multi-config builds, since my script assumes certain build directory layout (build/<config>/) that is not maintained by those specific builds. We support building for multiple configurations already through the script with single-config generators and in case somebody would have a need for multi-config generator they might just make a couple of extra steps to set up project manually.

riebl commented on tools/build.py:
```diff
@@ -103,9 +102,20 @@ def configure(self: 'Routines') -> None:
+                '-B', str(self._params.build_directory),
+                '-S', str(pathlib.Path.cwd()),
+                self._decorate_cmake_variable('CMAKE_EXPORT_COMPILE_COMMANDS', 'ON', 'BOOL'),
+                self._decorate_cmake_variable('CMAKE_CONFIGURATION_TYPES', ';'.join(self._params.build_configs))
```
I would not set this variable explicitly if the user has not explicitly set configuration types on the command line.

riebl commented on tools/build.py:
```diff
@@ -212,6 +233,11 @@ def parse_cli_args() -> argparse.Namespace:
     parser.add_argument('--profile', action='store', dest='profile',
         help='specify path to profile, or its name if available.'
     )
+    parser.add_argument('--generator-is-multiconfig', action='store_true', dest='is_generator_multiconfig',
```
The best CLI argument is the one you don't need ;-)

riebl commented on tools/run-artery.py:
```diff
@@ -0,0 +1,77 @@
+from pathlib import Path
+
+
+for template, value in copy.copy(globals()).items():
```
I guess these lines can be removed now

riebl commented on tools/run-artery.py:
```diff
@@ -0,0 +1,77 @@
+    cmd = []
+    working_directory = Path.cwd()
+
+    if args.artery_launch_conf is None:
```
Mark the option as required in its add_argument setup call.

riebl commented on tools/run-artery.py:
```diff
@@ -0,0 +1,77 @@
+    parser.add_argument('--all', action='store_true', help='sets runall mode')
+    parser.add_argument('-b', '--batchsize', dest='batch', action='store')
+    parser.add_argument('-j', '--jobs', dest='jobs', action='store')
+    parser.add_argument('--artery-launch-conf', action='store')
```
No need to prefix the command line option with "artery" in a script called run-artery :-)

AshFungor replied:
I just wish we won't overlap with any opp_run cli arguments in the future :)

riebl commented on tools/run-artery.py:
```diff
@@ -0,0 +1,77 @@
+
+
+def main():
+    global OPP_RUNALL, OPP_RUN, OPP_DEBUG_RUN, \
```
remove

AshFungor commented on .flake8:
```diff
@@ -0,0 +1,18 @@
+    __init__.py
+
+ignore =
+    F401
```
I adjusted contents to be more reasonable :)
We may change them later if needed, anyway :)
